File: backend/app/services/script_runner.py
"""脚本执行器 - 安全地执行用户脚本"""

import os
import subprocess
import logging
from typing import Optional, Callable
from dataclasses import dataclass, field
from pathlib import Path

from app.utils.command_validator import (
    validate_command,
    sanitize_env_vars,
    CommandValidationError,
    ALLOWED_INTERPRETERS
)

logger = logging.getLogger(__name__)


@dataclass
class ScriptResult:
    """脚本执行结果"""
    returncode: int
    output: str
    process: Optional[subprocess.Popen] = None


def run_command(
    command: str,
    scripts_dir: str,
    timeout: int = 300,
    env_vars: Optional[dict] = None,
    max_log_size: int = 10 * 1024 * 1024,
    on_output: Optional[Callable] = None,
) -> ScriptResult:
    """安全执行命令（支持完整命令格式，如 "python script.py"）

    Args:
        command: 完整命令，如 "python script.py" 或 "node app.js"
        scripts_dir: 脚本根目录
        timeout: 超时时间（秒）
        env_vars: 注入的环境变量
        max_log_size: 最大日志大小（字节）
        on_output: 逐行输出回调函数，签名 (line: str) -> None

    Returns:
        ScriptResult: 执行结果

    Raises:
        ValueError: 参数非法
        FileNotFoundError: 脚本文件不存在
    """
    # 1. 使用安全验证器验证命令
    try:
        interpreter, full_path = validate_command(command, scripts_dir)
    except CommandValidationError as e:
        error_msg = f"命令验证失败: {str(e)}\n"
        logger.error(error_msg)
        if on_output:
            on_output(error_msg)
        return ScriptResult(returncode=1, output=error_msg)

    # 2. 构建执行命令 (使用参数列表，不使用 shell=True)
    if interpreter == "python" or interpreter == "python3":
        # Python 添加 -u 参数禁用缓冲，实现实时输出
        cmd = [interpreter, "-u", full_path]
    elif interpreter == "ts-node":
        cmd = ["npx", "ts-node", full_path]
    else:
        cmd = [interpreter, full_path]

    # 3. 构建安全的环境变量
    env = {}
    # 保留必要的系统环境变量
    for key in ['PATH', 'SYSTEMROOT', 'PATHEXT', 'TEMP', 'TMP', 'HOME', 'USER']:
        if key in os.environ:
            env[key] = os.environ[key]

    # 添加传入的环境变量（经过清理）
    if env_vars:
        safe_env = sanitize_env_vars(env_vars)
        env.update(safe_env)

    logger.info(f"执行命令: {' '.join(cmd)}")

    try:
        # 4. 使用 subprocess.Popen 执行命令（不使用 shell=True）
        process = subprocess.Popen(
            cmd,  # 参数列表形式，防止 shell 注入
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            env=env,
            cwd=scripts_dir,
            encoding='utf-8',
            errors='replace',
            bufsize=1,
            universal_newlines=True,
            # 关键：不使用 shell=True
        )

        # 5. 读取输出，限制最大大小
        output_chunks = []
        total_size = 0
        truncated = False

        for line in iter(process.stdout.readline, ''):
            if not line:
                break
            total_size += len(line)
            if total_size > max_log_size:
                truncated = True
                break
            output_chunks.append(line)
            if on_output:
                try:
                    on_output(line)
                except Exception:
                    pass

        # 6. 等待进程结束，处理超时
        try:
            process.wait(timeout=timeout)
        except subprocess.TimeoutExpired:
            logger.warning(f"脚本执行超时 ({timeout}s)，正在终止进程...")
            process.terminate()
            try:
                process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                logger.error("进程终止失败，强制杀死进程")
                process.kill()
                process.wait()
            output_chunks.append(f"\n[超时] 脚本执行超过 {timeout} 秒，已强制终止\n")

        output = "".join(output_chunks)
        if truncated:
            output += f"\n[截断] 日志输出超过 {max_log_size // 1024 // 1024}MB，已自动截断\n"

        return ScriptResult(
            returncode=process.returncode if process.returncode is not None else 1,
            output=output,
            process=process,
        )

    except FileNotFoundError:
        error_msg = f"运行环境未安装: {interpreter}\n"
        logger.error(error_msg)
        if on_output:
            on_output(error_msg)
        return ScriptResult(returncode=1, output=error_msg)
    except Exception as e:
        error_msg = f"脚本执行异常: {str(e)}\n"
        logger.error(error_msg, exc_info=True)
        if on_output:
            on_output(error_msg)
        return ScriptResult(returncode=1, output=error_msg)

File: server/service/script_runner.go
package service

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

type ScriptResult struct {
	ReturnCode int
	Output     string
	Truncated  bool
}

type OnOutputFunc func(line string)

type OnProcessStartFunc func(process *os.Process)

func RunCommand(command, scriptsDir string, timeout int, envVars map[string]string, maxLogSize int, onOutput OnOutputFunc, onProcessStart ...OnProcessStartFunc) (*ScriptResult, *os.Process, error) {
	interpreter, scriptPath, err := validateCommand(command, scriptsDir)
	if err != nil {
		return nil, nil, err
	}

	cmd := buildCmd(interpreter, scriptPath, scriptsDir, envVars)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, nil, fmt.Errorf("failed to create stdout pipe: %w", err)
	}
	cmd.Stderr = cmd.Stdout

	if err := cmd.Start(); err != nil {
		return nil, nil, fmt.Errorf("failed to start process: %w", err)
	}

	process := cmd.Process

	if len(onProcessStart) > 0 && onProcessStart[0] != nil {
		onProcessStart[0](process)
	}

	var outputBuilder strings.Builder
	totalSize := 0
	truncated := false

	scanner := bufio.NewScanner(stdout)
	scanner.Buffer(make([]byte, 64*1024), 1024*1024)

	done := make(chan struct{})
	go func() {
		for scanner.Scan() {
			line := scanner.Text()
			if totalSize < maxLogSize {
				outputBuilder.WriteString(line + "\n")
				totalSize += len(line) + 1
				if onOutput != nil {
					onOutput(line)
				}
			} else if !truncated {
				truncated = true
				msg := "\n[日志已截断，超过最大大小限制]"
				outputBuilder.WriteString(msg)
				if onOutput != nil {
					onOutput(msg)
				}
			}
		}
		close(done)
	}()

	timer := time.NewTimer(time.Duration(timeout) * time.Second)
	defer timer.Stop()

	waitCh := make(chan error, 1)
	go func() {
		waitCh <- cmd.Wait()
	}()

	var returnCode int
	select {
	case err := <-waitCh:
		<-done
		if err != nil {
			if exitErr, ok := err.(*exec.ExitError); ok {
				returnCode = exitErr.ExitCode()
			} else {
				returnCode = 1
			}
		}
	case <-timer.C:
		KillProcessGroup(cmd.Process)
		<-done
		<-waitCh
		returnCode = -1
		msg := fmt.Sprintf("\n[任务超时，已在 %d 秒后终止]", timeout)
		outputBuilder.WriteString(msg)
		if onOutput != nil {
			onOutput(msg)
		}
	}

	return &ScriptResult{
		ReturnCode: returnCode,
		Output:     outputBuilder.String(),
		Truncated:  truncated,
	}, process, nil
}

var extInterpreterMap = map[string]string{
	".py": "python3",
	".js": "node",
	".ts": "ts-node",
	".sh": "bash",
}

var desiInterpreterMap = map[string]string{
	".js": "node",
	".py": "python3",
	".ts": "ts-node",
	".sh": "bash",
}

func validateCommand(command, scriptsDir string) (string, string, error) {
	parts := strings.Fields(command)
	if len(parts) < 2 {
		return "", "", fmt.Errorf("命令格式无效，格式: <解释器> <脚本路径>")
	}

	interpreter := parts[0]
	scriptPath := strings.Join(parts[1:], " ")

	if interpreter == "task" {
		ext := strings.ToLower(filepath.Ext(scriptPath))
		mapped, ok := extInterpreterMap[ext]
		if !ok {
			return "", "", fmt.Errorf("task 命令不支持的文件扩展名: %s", ext)
		}
		interpreter = mapped
	}

	if interpreter == "desi" {
		ext := strings.ToLower(filepath.Ext(scriptPath))
		mapped, ok := desiInterpreterMap[ext]
		if !ok {
			return "", "", fmt.Errorf("desi 命令不支持的文件扩展名: %s", ext)
		}
		interpreter = mapped
	}

	allowedInterpreters := map[string]bool{
		"python": true, "python3": true, "node": true, "ts-node": true, "bash": true,
	}
	if !allowedInterpreters[interpreter] {
		return "", "", fmt.Errorf("不支持的解释器: %s", interpreter)
	}

	dangerous := []string{"..", "~", "$", "`", ";", "|", "&", ">", "<"}
	for _, d := range dangerous {
		if strings.Contains(scriptPath, d) {
			return "", "", fmt.Errorf("脚本路径包含危险字符: %s", d)
		}
	}

	allowedExts := map[string]bool{
		".py": true, ".js": true, ".ts": true, ".sh": true,
	}
	ext := filepath.Ext(scriptPath)
	if !allowedExts[ext] {
		return "", "", fmt.Errorf("不支持的文件扩展名: %s", ext)
	}

	fullPath := filepath.Join(scriptsDir, scriptPath)
	fullPath, err := filepath.Abs(fullPath)
	if err != nil {
		return "", "", fmt.Errorf("无效路径: %w", err)
	}

	absScriptsDir, _ := filepath.Abs(scriptsDir)
	if !strings.HasPrefix(fullPath, absScriptsDir) {
		return "", "", fmt.Errorf("检测到路径遍历攻击")
	}

	if _, err := os.Stat(fullPath); os.IsNotExist(err) {
		return "", "", fmt.Errorf("脚本不存在: %s", scriptPath)
	}

	return interpreter, fullPath, nil
}

func buildCmd(interpreter, fullPath, scriptsDir string, envVars map[string]string) *exec.Cmd {
	var cmd *exec.Cmd

	switch interpreter {
	case "python", "python3":
		cmd = exec.Command(interpreter, "-u", fullPath)
	case "ts-node":
		cmd = exec.Command("npx", "ts-node", fullPath)
	default:
		cmd = exec.Command(interpreter, fullPath)
	}

	cmd.Dir = filepath.Dir(fullPath)
	cmd.Env = buildEnv(envVars)

	setPgid(cmd)

	return cmd
}

func buildEnv(envVars map[string]string) []string {
	safeKeys := []string{"PATH", "HOME", "USER", "LANG", "LC_ALL", "TZ"}
	if runtime.GOOS == "windows" {
		safeKeys = append(safeKeys, "SYSTEMROOT", "PATHEXT", "TEMP", "TMP", "APPDATA", "LOCALAPPDATA", "USERPROFILE")
	}

	env := make([]string, 0)
	for _, key := range safeKeys {
		if val := os.Getenv(key); val != "" {
			env = append(env, key+"="+val)
		}
	}

	dangerousVars := map[string]bool{
		"LD_PRELOAD": true, "LD_LIBRARY_PATH": true, "DYLD_INSERT_LIBRARIES": true,
	}

	for k, v := range envVars {
		if dangerousVars[k] {
			continue
		}
		if strings.ContainsRune(v, 0) {
			continue
		}
		env = append(env, k+"="+v)
	}

	return env
}

func RunInlineScript(content, scriptsDir string, envVars map[string]string, timeout int, onOutput OnOutputFunc) error {
	tmpFile := filepath.Join(scriptsDir, fmt.Sprintf(".hook_%d.sh", time.Now().UnixNano()))
	if err := os.WriteFile(tmpFile, []byte(content), 0755); err != nil {
		return err
	}
	defer os.Remove(tmpFile)

	cmd := exec.Command("bash", tmpFile)
	cmd.Dir = scriptsDir
	cmd.Env = buildEnv(envVars)
	setPgid(cmd)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return err
	}
	cmd.Stderr = cmd.Stdout

	if err := cmd.Start(); err != nil {
		return err
	}

	scanner := bufio.NewScanner(stdout)
	go func() {
		for scanner.Scan() {
			if onOutput != nil {
				onOutput(scanner.Text())
			}
		}
	}()

	timer := time.NewTimer(time.Duration(timeout) * time.Second)
	defer timer.Stop()

	waitCh := make(chan error, 1)
	go func() {
		waitCh <- cmd.Wait()
	}()

	select {
	case err := <-waitCh:
		return err
	case <-timer.C:
		KillProcessGroup(cmd.Process)
		<-waitCh
		return fmt.Errorf("钩子脚本超时，已超过 %d 秒", timeout)
	}
}

func RunHookScript(scriptName, scriptsDir string, envVars map[string]string, onOutput OnOutputFunc) {
	hookPath := filepath.Join(scriptsDir, scriptName)
	if _, err := os.Stat(hookPath); os.IsNotExist(err) {
		return
	}

	absPath, _ := filepath.Abs(hookPath)
	absDir, _ := filepath.Abs(scriptsDir)
	if !strings.HasPrefix(absPath, absDir) {
		return
	}

	cmd := exec.Command("bash", hookPath)
	cmd.Dir = scriptsDir
	cmd.Env = buildEnv(envVars)
	setPgid(cmd)

	stdout, _ := cmd.StdoutPipe()
	cmd.Stderr = cmd.Stdout

	if err := cmd.Start(); err != nil {
		if onOutput != nil {
			onOutput(fmt.Sprintf("[hook %s failed to start: %s]", scriptName, err))
		}
		return
	}

	scanner := bufio.NewScanner(stdout)
	go func() {
		for scanner.Scan() {
			if onOutput != nil {
				onOutput(scanner.Text())
			}
		}
	}()

	timer := time.NewTimer(60 * time.Second)
	defer timer.Stop()

	waitCh := make(chan error, 1)
	go func() {
		waitCh <- cmd.Wait()
	}()

	select {
	case <-waitCh:
	case <-timer.C:
		KillProcessGroup(cmd.Process)
		<-waitCh
	}
}
